File: src/main/java/no/nav/syfo/services/PartnerService.java
package no.nav.syfo.services;

import lombok.extern.slf4j.Slf4j;
import no.nav.syfo.consumer.syfopartnerinfo.PartnerInfoResponse;
import no.nav.syfo.consumer.syfopartnerinfo.SyfoPartnerInfoConsumer;
import no.nav.syfo.domain.Fastlege;
import no.nav.syfo.domain.Partnerinformasjon;
import no.nav.syfo.services.exceptions.PartnerinformasjonIkkeFunnet;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;


@Slf4j
@Service
public class PartnerService {

    private final SyfoPartnerInfoConsumer syfoPartnerInfoConsumer;

    public PartnerService(final SyfoPartnerInfoConsumer syfoPartnerInfoConsumer) {
        this.syfoPartnerInfoConsumer = syfoPartnerInfoConsumer;
    }

    Partnerinformasjon getPartnerinformasjon(Fastlege fastlege) {
        try {
            String fastlegeForeldreEnhetHerId = Optional.ofNullable(fastlege.foreldreEnhetHerId)
                    .orElseThrow(() -> new PartnerinformasjonIkkeFunnet("Kunne ikke finne partnerinformasjon! Fant ikke fastlegeForeldreEnhetHerId."))
                    .toString();
            PartnerInfoResponse partnerInfoResponse = getPartnerInfoResponse(fastlegeForeldreEnhetHerId)
                    .orElseThrow(() -> new PartnerinformasjonIkkeFunnet("Kunne ikke finne partnerinformasjon! Ingen partnerInfo er knyttet til fastlegeForeldreEnhetHerId."));
            return new Partnerinformasjon(String.valueOf(partnerInfoResponse.getPartnerId()), fastlegeForeldreEnhetHerId);
        } catch (PartnerinformasjonIkkeFunnet e) {
            throw e;
        } catch (Exception e) {
            log.warn("DIALOGMELDING-TRACE: Annen feil oppstått", e);
            throw e;
        }
    }

    private Optional<PartnerInfoResponse> getPartnerInfoResponse(String herId) {
        List<PartnerInfoResponse> partnerInfoResponses = syfoPartnerInfoConsumer.getPartnerId(herId);
        return Optional.ofNullable(partnerInfoResponses.isEmpty() ? null : partnerInfoResponses.get(0));
    }
}

File: src/main/kotlin/no/nav/syfo/dialogmelding/PartnerService.kt
package no.nav.syfo.dialogmelding

import no.nav.syfo.consumer.syfopartnerinfo.SyfoPartnerInfoConsumer
import no.nav.syfo.dialogmelding.exception.PartnerinformasjonIkkeFunnet
import no.nav.syfo.fastlege.domain.Fastlege
import no.nav.syfo.fastlege.domain.Partnerinformasjon
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class PartnerService(
    private val syfoPartnerInfoConsumer: SyfoPartnerInfoConsumer,
) {
    fun getPartnerinformasjon(fastlege: Fastlege): Partnerinformasjon {
        return try {
            val fastlegeForeldreEnhetHerId = fastlege.foreldreEnhetHerId
                ?: throw PartnerinformasjonIkkeFunnet("Kunne ikke finne partnerinformasjon! Fant ikke fastlegeForeldreEnhetHerId.")
            val partnerId = syfoPartnerInfoConsumer.getPartnerId(fastlegeForeldreEnhetHerId).firstOrNull()?.partnerId
                ?: throw PartnerinformasjonIkkeFunnet("Kunne ikke finne partnerinformasjon! Ingen partnerInfo er knyttet til fastlegeForeldreEnhetHerId.")
            Partnerinformasjon(
                partnerId = partnerId,
                herId = fastlegeForeldreEnhetHerId,
            )
        } catch (e: PartnerinformasjonIkkeFunnet) {
            throw e
        } catch (e: Exception) {
            log.warn("DIALOGMELDING-TRACE: Annen feil oppstått", e)
            throw e
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(PartnerService::class.java)
    }
}
